### ISSUE TITLE
[BUG] HEPv3 siptrace to Homer v7 - "Trailing stray characters"

### ISSUE THREAD

**Issue by solarmon** [OpenSIPS/opensips#2146]
**OpenSIPS version you are running**

```
version: opensips 2.4.5 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 3a566f1
main.c compiled on 09:53:43 Mar 15 2019 with gcc 4.8.5
```

**Describe the bug**

I'm trying to set up siptrace to send HEPv3 packets to a Homer 7 setup.

Currently this is not working and it seems the HEPv3 packets coming from opensips is not being ingested for some reason. Using hepgen (https://github.com/sipcapture/hepgen.js) to generate test HEPv3 traffic on the same opensips nodes is successful.

When comparing both he captured opensips and hepgen generated HEPv3 packets I notice the following difference which might explain why the opensips HEPv3 packets are not being ingested.

I'm using the HEP dissector plugin for Wireshark from:

https://github.com/sipcapture/hep-wireshark

The opensips HEPv3 packet is decoded as HEPv3 by the dissector plugin, but it gives a warning about "Trailing stray characters" for the packet. Inspecting the packet further I see that there is indeed extra data at the end. For example, for the HEPv3 packet for INVITE there is the following data:

```
.....70110F001-0000FA9B-5EEA5D1C-0000000E@ip.address
```

(IP address sanitised)

This seems to be the same value as for the Call-ID header:

```
Call-ID: 0110F001-0000FA9B-5EEA5D1C-0000000E@ip.address
```

I have the following config, which I pulled together from various sources - so I'm not sure whether this is correct or causing this issue:

```
### HEP Capture

#TCP Hep listener
listen=hep_tcp:w.x.y.z:6060
#UDP Hep listener
listen=hep_udp:w.x.y.z:6060

loadmodule "proto_hep.so"
modparam("proto_hep", "hep_id", "[homer]a.b.c.d:9060;transport=tcp;version=3;")

loadmodule "siptrace.so"
modparam("siptrace", "trace_on", 1)
modparam("siptrace", "trace_id", "[traceid]uri=hep:homer")

###
```

and I'm call sip_trace() withing the main route{} function (just for testing at the moment):

```
sip_trace("traceid","M","sip");
```
Why is opensips adding this extra data at the end of the HEPv3 packets and is this expected/normal?

If not expected/normal how should I change the opensips config to resolve it.

**To Reproduce**

See setup and config described above.

**Expected behavior**

No extraneous data at end of HEPv3 packet

**Relevant System Logs**

**OS/environment information**
 - Operating System: 
 - OpenSIPS installation: 
 - other relevant information:
```
CentOS Linux release 7.6.1810 (Core)
```

**Additional context**

**Comment by bogdan-iancu** [comment]
@solarmon , could you upgrade to the latest 2.4 version (git  or nightly builds) to double check if the problem is still there ?

**Comment by solarmon** [comment]
@bogdan-iancu 

I have put off updating as I wasn't sure of the correct process. I had built it from git source, but I'm not sure of the correct 'upgrade' process. 

Please can you advise of the appropriate upgrade process to make sure the current configuration is retained. This is being tested in a Dev environment so it is not impacting any production services, but I need to make sure there is a suitable upgrade process for when it is applied to production.

Thank you.

**Comment by solarmon** [comment]
@bogdan-iancu 

Just to give some context, I had used the following process to install from git source. I downloaded the source with:

git clone -b 2.4 https://github.com/OpenSIPS/opensips.git

Then used 'make menuconfig' to select modules for for compiling.

Should I use the same process, or is there a recommended 'upgrade' process to use instead?

Thank you.

**Comment by bogdan-iancu** [comment]
@solarmon , you go back to the original git clone you had from the first installation and run `git pull --rebase` to update. Or as you suggested, you can a fresh clone. After that just did what you did during the first installation

**Comment by solarmon** [comment]
@bogdan-iancu 

Thanks. I just did just a "git pull" and that seems to have updated my git clone. But I'll use your suggested method next time.

I have compiled and installed opensips and it is now at version 2.4.7:

```
version: opensips 2.4.7 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: bd49bfe
main.c compiled on 12:24:03 Jun 23 2020 with gcc 4.8.5
```

Unfortunately, I'm am still seeing the same 'Trailing stray characters' issue and the same extraneous "Call-ID" data at the end of the HEPv3 encapsulated SIP packet.

What do you suggest now?

**Comment by solarmon** [comment]
@bogdan-iancu 

I've been comparing the hepgen packets and the opensips sipcature packets and it seems there is a difference , which relates to this Call-ID value that I'm seeing.

In the hepgen HEPv3 packet, the "**Correlation ID**" section, which seems to also have the Call-ID value for it, is before the "**Encapsulated Payload [truncated]**" section.

In the opensips sipcature HEPv3 packet I see that the "**Correlation ID**" is **AFTER** "**Encapsulated" Payload [truncated]**", so appears after the encapsulated SIP packet.

I think this is the reason why the Wireshark HEPv3 dissector is warning and showing this Call-ID / Correlation ID after the SIP packet; and possibly why Homer/Heplify is rejecting it because it is malformed?

**Comment by solarmon** [comment]
@bogdan-iancu 

I think this "Trailing stray characters" and the "Correlation ID" and "Encapsulated" Payload [truncated]" ordering is a red herring, although I'd like to understand why the HEP dissector for Wireshark is giving such a warning.

I was able to use sngrep to send captured SIP packets on the opensips server and send it to my Homer/Heplify server and it appeared OK.

I noticed that both hepgen and sngrep were send HEPv3 packets as UDP and not TCP. I changed opensips to send HEPv3 packets in UDP and it work!

I thought the Homer docker setup was listening on both TCP and UDP port 9060, which is what is suggested by netstat.

I'm not sure why Homer/Heplify is not accepting HEPv3 TCP packets - I have a ticket open on their Github to ask this question.

**Comment by bogdan-iancu** [comment]
Hi @solarmon , thanks for the update here. I'm running several OpenSIPS 2.4 with HEP capturing and it works, the packets are displayed in HOMER - so the problem may be with the dissector, not with the actual packet.
And I did also run into problems with Heplify server listening on TCP - I failed to make it work do I'm using UDP for HEP (which works great).
I think we are good to close here, as there is no actual problem on the OpenSIPS side.

BTW, if you get a resolution on the HEP + TCP report at homer project, just update us here, it will be good to know, thank you
